### PR TITLE
Cleanup old IE/edge properties and implement unsupported browser error page

### LINKED ISSUE #4497: Browser update notification for outdated browsers
Ref https://github.com/owncloud/core/pull/26967

### PR DESCRIPTION
**Pull request by skjnldsv**
The message gets narrowed a bit, we only show recommendations for mobile apps if you're on a mobile and for sektop browsers if you're on a desktop machine.

| Desktop | Mobile |
|:---------:|:------:|
|dev skjnldsv com_unsupported|dev skjnldsv com_unsupported (1)|

Ref #20855

### CODE REVIEW

**Review by jancborchardt — APPROVED** [review]
Looks good! :)

Do you think we should offer some additional info like which browser wpuld work? But then we also need to specify version numbers and it gets nerdy fast.

So I think it's fine like you did it! :)

**Comment by skjnldsv** [comment]
> Do you think we should offer some additional info like which browser wpuld work? But then we also need to specify version numbers and it gets nerdy fast.

Yes, that was my question and why I asked for your advice :)
Maybe I can automatically build a list of supported browsers based on our configs :thinking:

**Comment by skjnldsv** [comment]
So, I can build a check based on our browserlist, but it's not feature-aware.

So some old FF versions might still be comptible with what we use, but using this method will redirect everyone that is not STRICTLY using one of the browser that is within our browserlist config into this "unsupported" page.

Anyone have a better idea? Maybe just show a warning and only redirect IE?
Or all dead browsers? 

cc @juliushaertl @PVince81 @ChristophWurst @eneiluj @danxuliu

File: .eslintrc.js [hunk 1]
```diff
@@ -9,6 +9,8 @@ module.exports = {
 		oc_userconfig: true,
 		dayNames: true,
 		firstDay: true,
+		supportedBrowsers: true,
```
Comment (ChristophWurst) [inline .eslintrc.js:12]: where is this global variable coming from?

Comment (skjnldsv) [reply .eslintrc.js:12]: Generated on webpack config from our browserlist config 
https://github.com/nextcloud/server/blob/11db32743e9cf785e7cc958436ac17296c4fd0de/webpack.common.js#L133

Comment (ChristophWurst) [reply .eslintrc.js:12]: Can't you just import the module wherever you use the global var?

Comment (skjnldsv) [reply .eslintrc.js:12]: maybe :thinking:

Comment (ChristophWurst) [reply .eslintrc.js:12]: That's fine. Webpack should still be able to link and inline the code.

Comment (skjnldsv) [reply .eslintrc.js:12]: Yeah, it also messes with which dependency is dev or not then, I prefer having this on buildtime than runtime

Comment (ChristophWurst) [reply .eslintrc.js:12]: But it *is* a runtime dependency. You depend on it to run the application.

Regardless of this I can't stress enough to pay attention when we use global variables. They are some of the basic anti patterns in software design. They aren't always avoidable in js, so let's use them wisely :pray:

Comment (skjnldsv) [reply .eslintrc.js:12]: The config is not though, the config is only used by babel.
But I agree with you :thinking:

File: core/src/views/UnsupportedBrowser.vue [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+				return this.t('core', '{name} version {version} and above', {
+					name, version,
+				})
+			})
+		},
+	},
+
+	beforeMount() {
+		// If the ignore token is set, redirect
+		if (browserStorage.getItem(browserStorageKey) === 'true') {
```
Comment (PVince81) [inline core/src/views/UnsupportedBrowser.vue:93]: do this earlier, like in main-unsupported-browser.js or even earlier ? this way no need to initialize a Vue instance just for redirecting afterwards.

File: package.json [hunk 3]
```diff
@@ -105,6 +105,9 @@
     "babel-jest": "^26.6.3",
     "babel-loader": "^8.2.2",
     "babel-loader-exclude-node-modules-except": "^1.1.2",
+    "browserslist-useragent-regexp": "^3.0.0",
+    "caniuse-db": "^1.0.30001237",
+    "chai": "^4.3.4",
```
Comment (PVince81) [inline package.json:110]: why do we need this test/assertion library out of a sudden ? I didn't see any test being adjusted

**Review by PVince81 — COMMENTED** [review]
see comments

File: package.json (hunk 3 above)
Comment (skjnldsv) [reply package.json:110]: rebase mess

File: .eslintrc.js (hunk 1 above)
Comment (skjnldsv) [reply .eslintrc.js]: Done, migrated to runtime!! :rocket:

File: core/src/views/UnsupportedBrowser.vue (hunk 2 above)
Comment (skjnldsv) [reply core/src/views/UnsupportedBrowser.vue]: This is only here when someone manually access this page despite having already allowed the browser. Which means a user had to manually specify this url.

The main redirect happens much sooner :wink:

Comment (PVince81) [reply core/src/views/UnsupportedBrowser.vue]: Ah okay, I missed that :+1:

**Comment by skjnldsv** [comment]
jsunit is not happy :thinking:

**Comment by skjnldsv** [comment]
Rebased and ready to go for 26 :rocket: 
- Updated screenshots
- Using our new style
- Fixed components
- Improved the UX a bit

**Review by artonge — APPROVED** [review]

**Comment by skjnldsv** [comment]
> I just wanted to test this but failed since testing was hard because Win11 removed Internet Explorer. I'll check back in a few days or latest next week when I have a Win10 device at my hands.

You can manually change your user agent too :)
https://developer.chrome.com/docs/devtools/device-mode/override-user-agent/

**Comment by szaimen** [comment]
> > I just wanted to test this but failed since testing was hard because Win11 removed Internet Explorer. I'll check back in a few days or latest next week when I have a Win10 device at my hands.
> 
> You can manually change your user agent too :) https://developer.chrome.com/docs/devtools/device-mode/override-user-agent/

good to know! Will check this out then :)

However I have the feeling that it is already so broken that the page doesnt even get shown on IE XD

**Review by szaimen — APPROVED** [review]
LGTM, though didnt review the code and wasnt able to check with a real IE. It is possible that the site doesnt load with that at all.

Though, should the items on the bottom not be centered?

File: core/src/views/UnsupportedBrowser.vue [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+	},
+}
+</script>
+
+<style lang="scss" scoped>
+.content-unsupported-browser {
+	display: flex;
+	justify-content: center;
+	width: 400px;
+	max-width: 90vw;
```
Comment (szaimen) [inline core/src/views/UnsupportedBrowser.vue:166]:
```suggestion
	max-width: 80vw;
```
loogs better on very small  widths

File: core/src/views/UnsupportedBrowser.vue [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+	&__continue {
+		display: block;
+		margin: 20px auto;
+	}
+
+	&__list {
+		margin-top: 60px;
+		margin-bottom: 30px;
+		li {
+			text-align: left;
```
Comment (szaimen) [inline core/src/views/UnsupportedBrowser.vue:186]: Maybe?

```suggestion
			text-align: center;
```

File: core/src/views/UnsupportedBrowser.vue [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+	display: flex;
+	justify-content: center;
+	width: 400px;
+	max-width: 90vw;
+	margin: auto;
+	padding: 30px;
+
+	.empty-content {
+		margin: 0;
+		&::v-deep .empty-content__icon {
```
Comment (Pytal) [inline core/src/views/UnsupportedBrowser.vue:172]:
```suggestion
		&:deep(.empty-content__icon) {
```
For stylelint https://github.com/nextcloud/stylelint-config/pull/65

File: core/src/views/UnsupportedBrowser.vue (hunk 5 above)
Comment (skjnldsv) [reply core/src/views/UnsupportedBrowser.vue:186]: No, it looks bad centered :)

File: core/src/views/UnsupportedBrowser.vue (hunk 6 above)
Comment (skjnldsv) [reply core/src/views/UnsupportedBrowser.vue:172]: I think we're too outdated on server and we can't use it for now, it doesn't translate for me, it stays `:deep` in the code

File: core/routes.php [hunk 7]
```diff
@@ -97,6 +99,9 @@
 
 		// Well known requests https://tools.ietf.org/html/rfc5785
 		['name' => 'WellKnown#handle', 'url' => '.well-known/{service}'],
+
+		// Unsupported browser
```
[the remark this answers is not in the document]
Comment (skjnldsv) [reply core/routes.php:103]: You're being picky :see_no_evil:

File: core/src/views/UnsupportedBrowser.vue (hunk 4 above)
Comment (skjnldsv) [reply core/src/views/UnsupportedBrowser.vue:166]: Ah right, the box-sizing is applied to content only! :see_no_evil:

File: core/src/views/UnsupportedBrowser.vue (hunk 6 above)
Comment (skjnldsv) [reply core/src/views/UnsupportedBrowser.vue]: image
Yep

File: config/config.sample.php [hunk 8]
```diff
@@ -2182,6 +2182,16 @@
 
 'login_form_autocomplete' => true,
 
+/**
+ * If your user is using an outdated browser, a warning will be shown
+ * to offer some guidance to upgrade and ensure a proper Nextcloud experience.
+ * They can still bypass it after they have read the warning.
+ *
+ * Simply set this property to "false", if you want to turn this feature off.
```
Comment (tcitworld) [inline config/config.sample.php:2190]: If I understand correctly.
```suggestion
 * Simply set this property to "true", if you want to turn this feature off.
```

Comment (skjnldsv) [reply config/config.sample.php:2190]: :sob: I just rebased everything!

Nice catch..... 🥲

Comment (skjnldsv) [reply config/config.sample.php]: Fixed

**Comment by AlvaroBrey** [comment]
I believe this change has broken the login flow in Android (maybe iOS too). See the following video:

<details>

https://user-images.githubusercontent.com/8677743/197213901-55448429-3386-44f8-b1ad-f7b4332a7d54.mp4

</details>

As seen in the video, the webview just goes to the Dashboard instead of completing the login flow.

The Android app uses the system webview for this, as a reference.

**Comment by AlvaroBrey** [comment]
> I believe this change may have broken the login flow in Android (maybe iOS too). See the following video:
> 
> As seen in the video, the webview just goes to the Dashboard instead of completing the login flow.
> 
> The Android app uses the Android system webview, as a reference.

Confirming, setting `no_unsupporting_browser_warning` to `true` lets the app login again

**Comment by szaimen** [comment]
> Confirming, setting `no_unsupporting_browser_warning` to `true` lets the app login again. Will investigate if we can work around it in the app somehow.

Another option wouold be adding the webflow user agent to the supported browsers... what is the user agent of such?

**Comment by AlvaroBrey** [comment]
After some testing, it looks like the unsupported browser page completely discards the initially loaded URL (in this case `/login/flow/...`) and redirects to `/index.php`.

> Another option wouold be adding the webflow user agent to the supported browsers... what is the user agent of such?

This seems difficult. We have two different ways of creating user agents:

- For most cases we use a [combination of properties from the Android device](https://github.com/nextcloud/android/blob/2e49782298323deb2e86c4aed66c3fa101cc8675/app/src/main/java/com/owncloud/android/authentication/AuthenticatorActivity.java#L339-L339). This would look like this for my emulator, for example: `Google sdk_gphone64_x86_64 (Android)`
- For one specific case I don't quite understand, we use a pre-configured string, which is the same we use for API requests. This would look like: `Mozilla/5.0 (Android) Nextcloud-android/3.23.0 Alpha1`,  BUT this is changed for branded builds.

**Comment by szaimen** [comment]
So what can we do here then? Improve the browser warning to redirect correctly?

**Comment by AlvaroBrey** [comment]
> So what can we do here then? Improve the browser warning to redirect correctly?

That is needed, yes. Because even if we change the Android app to use the default browser instead of a Webview, if the default browser is outdated the same problem will occur and users will not be able to login.

Another option is to disable the outdated warning for the `/login/flow` path and subpaths. Not sure if this is sensible or even possible.

**Comment by szaimen** [comment]
cc @skjnldsv on this then :)

**Comment by skjnldsv** [comment]
> > Confirming, setting `no_unsupporting_browser_warning` to `true` lets the app login again. Will investigate if we can work around it in the app somehow.
> 
> Another option wouold be adding the webflow user agent to the supported browsers... what is the user agent of such?

Are you setting something specific to the headers that we could use maybe?

> the same problem will occur and users will not be able to login

Well, it's a warning, they can bypass it.
But it's defeating the purpose since we're loging in for the auth flow only.
I guess we could check if this is a login for app authentication or just a standard login which the user aims to use the web ui :)

In any case, this si why we merge such things early, to catch them! Nice found @AlvaroBrey :tada:

**Comment by skjnldsv** [comment]
> * For most cases we use a [combination of properties from the Android device](https://github.com/nextcloud/android/blob/2e49782298323deb2e86c4aed66c3fa101cc8675/app/src/main/java/com/owncloud/android/authentication/AuthenticatorActivity.java#L339-L339). This would look like this for my emulator, for example: `Google sdk_gphone64_x86_64 (Android)`

We could add a bypass if the server as debug mode enabled too


> * For one specific case I don't quite understand, we use a pre-configured string, which is the same we use for API requests. This would look like: `Mozilla/5.0 (Android) Nextcloud-android/3.23.0 Alpha1`,  BUT this is changed for branded builds IIRC. @tobiasKaminsky can hopefully shed some light here

Yep, will need to know how the user agents behave for branded or not.
Otherwise I can just check for the Nextcloud word in the agent :)

**Comment by AlvaroBrey** [comment]
> Well, it's a warning, they can bypass it.

Right now the user cannot bypass it, as the login flow URL (with its included tokens) is lost on the redirection.

>  I guess we could check if this is a login for app authentication or just a standard login which the user aims to use the web ui :)

This is why I suggested to bypass the check on `/login/flow` (and probably `/login/v2/flow` too, though the Android app does not use that one), since those are only meant for client authentication, not for web use AFAIK.

> Are you setting something specific to the headers that we could use maybe?

Right now we're adding `OCS-APIREQUEST=true` to the first request only (`/index.php/login/flow`) but not to subsequent requests. If needed we can probably add more headers/try to add them on every request through the login process.

**Comment by skjnldsv** [comment]
> Right now the user cannot bypass it, as the login flow URL (with its included tokens) is lost on the redirection.

Good call, then we should keep the redirect too :)

> This is why I suggested to bypass the check on `/login/flow` (and probably `/login/v2/flow` too, though the Android app does not use that one), since those are only meant for client authentication, not for web use AFAIK.

I would like something a bit more standard than an url.
What if we have more URLs to cover ?

> Right now we're adding OCS-APIREQUEST=true to the first request only (/index.php/login/flow) but not to subsequent requests. If needed we can probably add more headers/try to add them on every request through the login process.

Let's see what I can do :)

**Comment by skjnldsv** [comment]
> This is why I suggested to bypass the check on `/login/flow` (and probably `/login/v2/flow` too, though the Android app does not use that one), since those are only meant for client authentication, not for web use AFAIK.

Maybe we could just load this on default template, and not login/maintenance/error/twofactor authentifications pages. Let's see what I can do here :)

**Comment by skjnldsv** [comment]
Fix in https://github.com/nextcloud/server/pull/34741

**Pull request merged**